x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/BinaryDenseVector.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.vectors.query;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.Version;
+import org.elasticsearch.xpack.vectors.mapper.VectorEncoderDecoder;
+
+import java.nio.ByteBuffer;
+
+public class BinaryDenseVector implements DenseVector {
+    protected final BytesRef docVector;
+    protected final int dims;
+    protected final Version indexVersion;
+
+    protected float[] decodedDocVector;
+
+    public BinaryDenseVector(BytesRef docVector, int dims, Version indexVersion) {
+        this.docVector = docVector;
+        this.indexVersion = indexVersion;
+        this.dims = dims;
+    }
+
+    @Override
+    public float[] getVector() {
+        if (decodedDocVector == null) {
+            decodedDocVector = new float[dims];
+            VectorEncoderDecoder.decodeDenseVector(docVector, decodedDocVector);
+        }
+        return decodedDocVector;
+    }
+
+    @Override
+    public float getMagnitude() {
+        return VectorEncoderDecoder.getMagnitude(indexVersion, docVector);
+    }
+
+    @Override
+    public double dotProduct(float[] queryVector) {
+        ByteBuffer byteBuffer = wrap(docVector);
+
+        double dotProduct = 0;
+        for (float v : queryVector) {
+            dotProduct += byteBuffer.getFloat() * v;
+        }
+        return dotProduct;
+    }
+
+    @Override
+    public double dotProduct(QueryVector queryVector) {
+        ByteBuffer byteBuffer = wrap(docVector);
+
+        double dotProduct = 0;
+        for (int i = 0; i < queryVector.size(); i++) {
+            dotProduct += byteBuffer.getFloat() * queryVector.get(i);
+        }
+        return dotProduct;
+    }
+
+    /**
+     * dotProduct of doc vector and query vector that normalizes the query vector while performing the calculation.
+     */
+    protected double dotProduct(QueryVector qv, float qvMagnitude) {
```

Comment:
I don't think we need these special methods that take `qvMagnitude`, because this calculation is the equivalent to just doing a  normal `dotProduct`, then dividing the result by `qvMagnitude`.

Reply:
Removed these methods and updated cosSim implementation.